akka-stream/src/main/scala/akka/stream/impl/Sinks.scala

```diff
@@ -380,3 +361,150 @@ private[akka] final class SinkQueueAdapter[T](delegate: SinkQueue[T]) extends ak
+
+  override def toString: String = "AdvancedPublisherSink"
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes) = {
+    import AdvancedPublisherSink._
+    val InputBuffer(initBuffer, maxBuffer) = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16))
+    require(maxBuffer > 0, "Max buffer size must be greater than 0")
+    require(initBuffer > 0, "Init buffer size must be greater than 0")
+
+    val stageLogic = new GraphStageLogic(shape) with CallbackWrapper[PublisherSinkMessages[T]] with SubscriberManagement[T] with InHandler {
+      private var downstreamDemand: Long = 0L
+      private var upstreamCompleted = false
+      private var failedWithError: Throwable = null
+
+      override type S = SubscriptionWithCursor[T]
+
+      override def maxBufferSize: Int = maxBuffer
+      override def initialBufferSize: Int = initBuffer
+
+      setHandler(in, this)
+
+      override def preStart(): Unit = {
+        setKeepGoing(true)
+        initCallback(callback.invoke)
+      }
+
+      override def postStop(): Unit = {
+        stopCallback {
+          case Subscribe(subscriber) ⇒
+            subscriber.onSubscribe(CancelledSubscription)
+            subscriber.onError(if (failedWithError == null) ActorPublisher.NormalShutdownReason else failedWithError)
+          case _ ⇒ //keep silence
+        }
+      }
+
+      private val callback: AsyncCallback[PublisherSinkMessages[T]] =
+        getAsyncCallback {
+          case Subscribe(subscriber) ⇒ strategy match {
+            case DrainIfNoSubscribtions if cursors.nonEmpty ⇒ rejectAdditionalSubscriber(subscriber, "Sink.asPublisher(fanout = false)")
+            case _ ⇒ registerSubscriber(subscriber.asInstanceOf[Subscriber[_ >: T]])
+          }
+          case RequestMore(subscriber, elements) ⇒ moreRequested(subscriber, elements)
+          case Cancel(subscription) ⇒
+            unregisterSubscription(subscription)
+            if (cursors.isEmpty && strategy == Fanout) completeStage()
+        }
+
+      override def onPush(): Unit = {
+        if (cursors.isEmpty) {
+          //drain if no subscriptions
+          downstreamDemand = 0
+          grab(in)
+          pull(in)
+        } else {
+          pushToDownstream(grab(in))
+          if (upstreamCompleted && cursors.isEmpty) completeStage() else pullIfNeeded()
+        }
+      }
+
+      private def pullIfNeeded(): Unit = {
+        if (downstreamDemand > 0 && !hasBeenPulled(in)) {
+          downstreamDemand -= 1
+          pull(in)
+        }
+      }
+
+      override def onUpstreamFinish(): Unit = {
+        completeDownstream()
+        upstreamCompleted = true
+        if (cursors.isEmpty) completeStage() else setKeepGoing(true)
+      }
+
+      override def onUpstreamFailure(ex: Throwable): Unit = {
+        failedWithError = ex
+        abortDownstream(ex)
+        failStage(ex)
+      }
+
+      override protected def shutdown(): Unit = completeStage()
```

Comment:
We call shutdown()= completeStage() only from getAsyncCallback {_} so it's safe.